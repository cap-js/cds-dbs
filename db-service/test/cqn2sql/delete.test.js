'use strict'
const cds = require('@sap/cds/lib')
const cqn2sql = require('../../lib/cqn2sql')

beforeAll(async () => {
  cds.model = await cds.load(__dirname + '/testModel').then(cds.linked)
})
describe('delete', () => {
  test('test with from entity', () => {
    const cqnDelete = DELETE.from(cds.model.definitions.Foo)
    const { sql } = cqn2sql(cqnDelete)
    expect(sql).toMatchSnapshot()
  })

  test('test with from ref', () => {
    const cqnDelete = {
      DELETE: {
        from: { ref: ['Foo'] },
      },
    }
    const { sql } = cqn2sql(cqnDelete)
    expect(sql).toMatchSnapshot()
  })

  test('test with from ref and alias', () => {
    const cqnDelete = {
      DELETE: {
        from: { ref: ['Foo'], as: 'lala' },
      },
    }
    const { sql } = cqn2sql(cqnDelete)
    expect(sql).toMatchSnapshot()
  })

  test('test with from string and where clause', () => {
    const cqnDelete = {
      DELETE: {
        from: 'Foo',
        where: [{ ref: ['x'] }, '<', { val: 9 }],
      },
    }
    const { sql } = cqn2sql(cqnDelete)
    expect(sql).toMatchSnapshot()
  })

  test("test simple cascade delete for entity with 'not in'", () => {
    const cqnDelete = {
      DELETE: {
        from: 'Foo',
        where: [
          {
            ref: ['x'],
          },
          'not in',
          {
            SELECT: {
              columns: [
                {
                  ref: ['a'],
                },
              ],
              from: { ref: ['Foo2'] },
            },
          },
        ],
      },
    }
    const { sql } = cqn2sql(cqnDelete)
    expect(sql).toMatchSnapshot()
  })

  test("test simple cascade delete for entity with 'not exists'", () => {
    const cqnDelete = {
      DELETE: {
        from: 'Books',
        where: [
          '(',
          {
            ref: ['author', 'id'],
          },
          'is not null',
          ')',
          'and',
          'not exists',
          {
            SELECT: {
              columns: [
                {
                  val: 1,
                  as: '_exists',
                },
              ],
              from: {
                ref: ['Author'],
              },
              where: [
                {
                  ref: ['id'],
                },
                '=',
                {
                  ref: ['parent', 'ID'],
                },
              ],
            },
          },
        ],
      },
    }
    const { sql } = cqn2sql(cqnDelete)
    expect(sql).toMatchSnapshot()
  })

  test("test complex cascade delete for entity with 'not exists'", () => {
    const cqnDelete = {
      DELETE: {
        from: 'Books',
        where: [
          '(',
          {
            ref: ['author', 'id'],
          },
          'is not null',
          'or',
          {
            ref: ['author', 'version'],
          },
          'is not null',
          ')',
          'and',
          'not exists',
          {
            SELECT: {
              columns: [
                {
                  val: 1,
                  as: '_exists',
                },
              ],
              from: {
                ref: ['Author'],
              },
              where: [
                {
                  ref: ['parent', 'ID'],
                },
                '=',
                {
                  ref: ['parent', 'code'],
                },
                'and',
                {
                  ref: ['parent', 'descr'],
                },
                '=',
                {
                  ref: ['Author', 'version'],
                },
              ],
            },
          },
        ],
      },
    }
    const { sql } = cqn2sql(cqnDelete)
    expect(sql).toMatchSnapshot()
  })
  // Do we need this test ?
  xtest("test complex reverse cascade delete for entity with 'exists'", () => {
    const cqnDelete = {
      DELETE: {
        from: 'sub1_1_1',
        where: [
          '(',
          {
            ref: ['sub1_1_1', 'sub1_1_parentForeignKey1_1_1'],
          },
          'is not null',
          'or',
          {
            ref: ['sub1_1_1', 'sub1_1_parentForeignVersion1_1_1'],
          },
          'is not null',
          ')',
          'and',
          'exists',
          {
            SELECT: {
              columns: [
                {
                  val: 1,
                  as: '_exists',
                },
              ],
              from: {
                ref: ['sub1_1'],
              },
              where: [
                {
                  ref: ['sub1_1_1', 'sub1_1_parentForeignKey1_1_1'],
                },
                '=',
                {
                  ref: ['sub1_1', 'key1_1'],
                },
                'and',
                {
                  ref: ['sub1_1_1', 'sub1_1_parentForeignVersion1_1_1'],
                },
                '=',
                {
                  ref: ['sub1_1', 'version1_1'],
                },
                'and',
                '(',
                '(',
                {
                  ref: ['sub1_1', 'sub1_parentForeignKey1_1'],
                },
                'is not null',
                'or',
                {
                  ref: ['sub1_1', 'sub1_parentForeignVersion1_1'],
                },
                'is not null',
                ')',
                'and',
                'exists',
                {
                  SELECT: {
                    columns: [
                      {
                        val: 1,
                        as: '_exists',
                      },
                    ],
                    from: {
                      ref: ['sub1'],
                    },
                    where: [
                      {
                        ref: ['sub1_1', 'sub1_parentForeignKey1_1'],
                      },
                      '=',
                      {
                        ref: ['sub1', 'key1'],
                      },
                      'and',
                      {
                        ref: ['sub1_1', 'sub1_parentForeignVersion1_1'],
                      },
                      '=',
                      {
                        ref: ['sub1', 'version1'],
                      },
                      'and',
                      '(',
                      '(',
                      {
                        ref: ['sub1', 'root_rootForeignKey1'],
                      },
                      'is not null',
                      'or',
                      {
                        ref: ['sub1', 'root_rootForeignVersion1'],
                      },
                      'is not null',
                      ')',
                      'and',
                      'exists',
                      {
                        SELECT: {
                          columns: [
                            {
                              val: 1,
                              as: '_exists',
                            },
                          ],
                          from: {
                            ref: ['root'],
                          },
                          where: [
                            {
                              ref: ['sub1', 'root_rootForeignKey1'],
                            },
                            '=',
                            {
                              ref: ['root', 'rootKey'],
                            },
                            'and',
                            {
                              ref: ['sub1', 'root_rootForeignVersion1'],
                            },
                            '=',
                            {
                              ref: ['root', 'rootVersion'],
                            },
                            'and',
                            '(',
                            {
                              ref: ['rootKey'],
                            },
                            '=',
                            {
                              val: 1,
                            },
                            'and',
                            {
                              ref: ['rootVersion'],
                            },
                            '=',
                            {
                              val: 'active',
                            },
                            ')',
                          ],
                        },
                      },
                      ')',
                    ],
                  },
                },
                ')',
              ],
            },
          },
        ],
      },
    }
    const expected = {
      sql: 'DELETE FROM sub1_1_1 WHERE ( sub1_1_1.sub1_1_parentForeignKey1_1_1 IS NOT NULL OR sub1_1_1.sub1_1_parentForeignVersion1_1_1 IS NOT NULL ) AND EXISTS ( SELECT 1 AS _exists FROM sub1_1 WHERE sub1_1_1.sub1_1_parentForeignKey1_1_1 = sub1_1.key1_1 AND sub1_1_1.sub1_1_parentForeignVersion1_1_1 = sub1_1.version1_1 AND ( ( sub1_1.sub1_parentForeignKey1_1 IS NOT NULL OR sub1_1.sub1_parentForeignVersion1_1 IS NOT NULL ) AND EXISTS ( SELECT 1 AS _exists FROM sub1 WHERE sub1_1.sub1_parentForeignKey1_1 = sub1.key1 AND sub1_1.sub1_parentForeignVersion1_1 = sub1.version1 AND ( ( sub1.root_rootForeignKey1 IS NOT NULL OR sub1.root_rootForeignVersion1 IS NOT NULL ) AND EXISTS ( SELECT 1 AS _exists FROM root WHERE sub1.root_rootForeignKey1 = root.rootKey AND sub1.root_rootForeignVersion1 = root.rootVersion AND ( rootKey = 1 AND rootVersion = ? ) ) ) ) ) )',
      values: ['active'],
    }
    expect(cqn2sql(cqnDelete)).toEqual(expected)
  })
})
