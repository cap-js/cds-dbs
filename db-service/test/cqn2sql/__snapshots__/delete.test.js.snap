// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete test complex cascade delete for entity with 'not exists' 1`] = `"DELETE FROM Books as Books WHERE ( Books.author_id is not null or Books.author_version is not null ) and not exists (SELECT 1 as _exists FROM Author as Author left JOIN Books as parent ON parent.ID = Author.parent_ID WHERE Author.parent_ID = parent.code and parent.descr = Author.version)"`;

exports[`delete test simple cascade delete for entity with 'not exists' 1`] = `"DELETE FROM Books as Books WHERE ( Books.author_id is not null ) and not exists (SELECT 1 as _exists FROM Author as Author WHERE Author.id = Author.parent_ID)"`;

exports[`delete test simple cascade delete for entity with 'not in' 1`] = `"DELETE FROM Foo as Foo WHERE Foo.x not in (SELECT Foo2.a FROM Foo2 as Foo2)"`;

exports[`delete test with from entity 1`] = `"DELETE FROM Foo as Foo"`;

exports[`delete test with from ref 1`] = `"DELETE FROM Foo as Foo"`;

exports[`delete test with from ref and alias 1`] = `"DELETE FROM Foo as lala"`;

exports[`delete test with from string and where clause 1`] = `"DELETE FROM Foo as Foo WHERE Foo.x < 9"`;
